src/components/implementation/no_interface/scheddev/sl.h

```diff
@@ -174,11 +175,177 @@ sl_cs_exit(void)
+ *	     +ve - number of cycles elapsed from abs_timeout before the thread
+ *		   was woken up by Timeout module.
+ */
+cycles_t sl_thd_block_timeout(thdid_t tid, cycles_t abs_timeout);
+/*
+ * blocks for on periodic wakeup based on task period.
+ *
+ * @returns: 0 if the thread is woken up by external events before timeout.
+ *           +ve - number of periods elapsed. (1 if it wokeup exactly at timeout = next period)
+ */
+unsigned sl_thd_block_periodic(thdid_t tid);
+int  sl_thd_block_no_cs(struct sl_thd *t, int is_timeout);
+/* wakeup a thread that has (or soon will) block */
+void sl_thd_wakeup(thdid_t tid);
+int  sl_thd_wakeup_no_cs(struct sl_thd *t);
+void sl_thd_yield(thdid_t tid);
+
+/* The entire thread allocation and free API */
+struct sl_thd *sl_thd_alloc(cos_thd_fn_t fn, void *data);
+struct sl_thd *sl_thd_comp_alloc(struct cos_defcompinfo *comp);
+void sl_thd_free(struct sl_thd *t);
+
+void sl_thd_param_set(struct sl_thd *t, sched_param_t sp);
+
 static inline cycles_t
 sl_now(void)
 { return ps_tsc(); }
 
 /*
+ * Time and timeout API.
+ *
+ * This can be used by the scheduler policy module *and* by the
+ * surrounding component code.  To avoid race conditions between
+ * reading the time, and setting a timeout, we avoid relative time
+ * measurements.  sl_now gives the current cycle count that is on an
+ * absolute timeline.  The periodic function sets a period that can be
+ * used when a timeout has happened, the relative function sets a
+ * timeout relative to now, and the oneshot timeout sets a timeout on
+ * the same absolute timeline as returned by sl_now.
+ */
+void sl_timeout_period(cycles_t period);
+
+static inline cycles_t
+sl_timeout_period_get(void)
+{ return sl__globals()->period; }
+
+static inline void
+sl_timeout_oneshot(cycles_t absolute_us)
+{
+	sl__globals()->timer_next   = absolute_us;
+	sl__globals()->timeout_next = tcap_cyc2time(absolute_us);
+}
+
+static inline void
+sl_timeout_relative(cycles_t offset)
+{ sl_timeout_oneshot(sl_now() + offset); }
+
+static inline microsec_t
+sl_cyc2usec(cycles_t cyc)
+{ return cyc / sl__globals()->cyc_per_usec; }
+
+static inline microsec_t
+sl_usec2cyc(microsec_t usec)
+{ return usec * sl__globals()->cyc_per_usec; }
+
+/* Timeout and wakeup functionality */
+#define SL_TIMEOUT_HEAP()       (&timeout_heap.h)
+
+struct timeout_heap {
+	struct heap  h;
+	void        *data[SL_MAX_NUM_THDS];
+} timeout_heap;
+
+/* wakeup any blocked threads! */
+static void
+sl_timeout_wakeup_expired(cycles_t now)
+{
+	if (!heap_size(SL_TIMEOUT_HEAP())) return;
+
+	do {
+		struct sl_thd *tp, *th;
+
+		tp = heap_peek(SL_TIMEOUT_HEAP());
+		assert(tp);
+
+		/* FIXME: logic for wraparound in current tsc */
+		if (tp->timeout_cycs > now) break;
+
+		th = heap_highest(SL_TIMEOUT_HEAP());
+		assert(th && th == tp);
+		th->timeout_idx = -1;
+
+		assert(th->wakeup_cycs == 0);
+		th->wakeup_cycs = now;
+		sl_thd_wakeup_no_cs(th);
```

Comment:
`sl_thd_block()` which is used for blocking forever (waiting for a wakeup event), does not add that thread to the timeout heap. Timeout heap only contains threads that have requested for a timeout wakeup. So it will not wakeup a thread that's blocked without timeout. 

I should however add an assert for checking the thread state is `BLOCKED_TIMEOUT`, that will also serve as documentation.